### ISSUE TITLE
Extending SciPy COO Sparse Arrays to N Dimensions

### ISSUE THREAD

**Issue by anushkasuyal** [scipy/scipy#21169]
### Summary
This project is undertaken as part of my summer internship at Quansight Labs, mentored by @izaid and @rgommers.
I would like to propose extending the current COO sparse array support in SciPy to n-dimensions. This enhancement targets users dealing with high-dimensional data, providing efficient storage and computation, reducing memory overhead, and avoiding unnecessary dense format computations. It will also be helpful throughout many other parts of SciPy that rely on sparse arrays. For instance, there are many methods that use sparse arrays to compute a result, and extending them to n dimensions will enable broadcasting via stacked sets of sparse arrays.

### Motivation
Currently, SciPy's sparse arrays are limited to 1D and 2D, with high-dimensional operations relying on dense NumPy arrays, leading to significant memory and computational inefficiencies. Extending COO sparse arrays to n dimensions will:

1. **Improve Efficiency:** Avoid dense format storage for high-dimensional sparse tensors.
2. **Enhance Performance:** Reduce unnecessary computations in matrix and tensor operations.
3. **Expand Functionality:** Enable direct manipulation of n-dimensional sparse data.

### Proposed Changes
The implementation can be approached in a phased manner, focusing on one chunk at a time. Community feedback will be sought at each stage to ensure alignment with user needs.

**1.	Data Structure and Basic Methods**

- Extend COO format to support creation of n-dimensional sparse arrays (`toarray`), and facilitate conversions between COO and dense formats.
- Implement/extend basic methods to n-dimensions.
- Implement element wise arithmetic operations (addition, subtraction etc. of COO arrays)

**2.	Matrix Multiplication and Tensor Operations**
- Extend matrix multiplication for n-dimensional COO arrays (i.e. sparse array alternative for `numpy.matmul` (may require modifying compiled code for efficiency).
- Implement basic tensor operations for sparse arrays like `tensordot` and `tensorsolve`.

**3.	Other operations**
- Adapt other existing linear algebra functions for n-dimensional COO arrays.
- Implement Boolean comparators for COO arrays (currently, most comparisons are routed via the CSR format, which wouldn’t support sparse arrays of dim>2)
- Ensure seamless integration with other SciPy modules.

### Implementation Plan
If the above workflow looks appropriate, I plan to have the first PR include the data structure and basic methods. Based on my exploration of the `_coo_base` class in _coo.py so far, I have noticed that quite a few methods have been defined in a generalised manner (perhaps keeping in mind extending to higher dimensions in the future), which allows for less modifications required. The main focus would be the methods that were only defined to work for 1D and 2D, and extending them to nD. This may require modifying compiled code for efficiency (such as having an n-dimensional version of the C++ function `coo_todense`).
Then, other basic methods such `reshape`, `count_nonzero`, `eliminate_zero`, `sum_duplicates`, `transpose` etc. could be tackled.
Once we have these working, arithmetic operations would be the next step. Currently, addition and subtraction of sparse arrays is routed via CSR for all formats (in the class `_spbase` in _base.py), just for consistency. This will, however, not work for higher dimensional COO arrays. It would then be suitable to have these methods declared separately for coo_array objects, for adding and subtracting sparse arrays of dimension>2.
This would potentially be sufficient implementation for the first PR. The next PR could then cover matrix multiplication and tensor operations.
I have so far spent 2 weeks on this, and have some confidence that this will work. There’s still quite a lot to do, but before proceeding further I thought this would be a good time for to get some feedback.

Here is the branch I've currently been working on - [extend-coo-nd](https://github.com/anushkasuyal/scipy/tree/extend-coo-nd)

**Comment by rgommers** [comment]
Cc @perimosocordiae @dschult who may have thoughts here (and commented on gh-21105).

**Comment by dschult** [comment]
Thanks for this work and set of ideas. 
I think it is natural to have multiple approaches to nD. Here is one I have been exploring. 

The idea is to lean heavily on the 2D code already in sparsetools and on NumPy's tools for index manipulation. I am hopeful this will give good performance (numpy is fast) without adding to the build time or library size (two issues I've heard concerns about from @rgommers).  It also seems good to take advantage of the numpy effort especially if it will ease some of the future burden for maintenance here.

For example: The `toarray()` function for COO nD arrays can be something like:
```python
    def toarray(self):
        flat_indices = np.ravel_multi_index(self.coords, self.shape)
        dense_result = np.zeros(self.shape, dtype=self.dtype)
        dense_result.flat[flat_indices] = self.data
        return dense_result
``` 
Even if this is 10-20% slower than a sparsetools version, I would argue that it is worth that performance gap for the sake of other benefits. Of course 100% slower means we should implement it in sparsetools unless there are other issues involved there.

In addition, I have a proof-of-concept class for a different approach to nD for n>2. The idea is to partially flatten the nD array coords to 2D. For example, for COO, we might flatten all coords except the last one into the rows of the 2D matrix and the last dimension would be the columns in 2D. Or (as in the POC version) the user can pick which axes to ravel into rows and the others go to columns.  The dimension reduction is done by `np.ravel_multi_index` and unraveling by `np.unravel_index`. 

In the POC the nD array stores the 2D array coords along with the nD coords and info about which axes go to the rows in 2D. But in the end, saving coords memory at the cost of repeated ravels might be better -- we'll see.  The key to this approach is two functions that convert between nD coords and 2D coords (with data in the same order). All the element-wise operations for 2D work seamlessly as demonstrated by the `__add__` method which adds the 2D versions of the arrays, calls `get_nd_coords` on the result and constructs a new nD coo_array from those coords.

Finally, this approach extends quite nicely to CSR format. We provide a list of axes to be compressed and the other remain uncompressed. Converting from COO involves compressing the 2D rep of the COO array with existing sparsetools methods. Similarly, `csr_array.tocoo()` would uncompress using sparsetools and then expand to nD using `np.unravel_index`.

A [gist of my proof-of-concept code is here](https://gist.github.com/dschult/6b65bd18cd425d07b1a659eb7f67bc58).

**Comment by rgommers** [comment]
> without adding to the build time or library size (two issues I've heard concerns about from @rgommers).

Good point, yes. @anushkasuyal when adding more C++ code, it'd be good to check binary size changes in release mode and build time. For a release build, you do a clean build with say `git clean -xdf && python dev.py build --release`. Build timing you can measure/visualize with `tools/ninjatracing.py`. Small increases are fine if there's a performance benefit to justify it, but we want to avoid doubling the size or build time of some already-heavy extension module.

**Comment by rgommers** [comment]
> Even if this is 10-20% slower than a sparsetools version, I would argue that it is worth that performance gap for the sake of other benefits. Of course 100% slower means we should implement it in sparsetools unless there are other issues involved there.

That sounds about right to me.

> In addition, I have a proof-of-concept class for a different approach to nD for n>2.

If there are two approached already here (say for `def toarray`), it may be worth comparing those now to better understand the tradeoff @anushkasuyal.

**Comment by anushkasuyal** [comment]
> The idea is to lean heavily on the 2D code already in sparsetools and on NumPy's tools for index manipulation. I am hopeful this will give good performance (numpy is fast) without adding to the build time or library size (two issues I've heard concerns about from @rgommers). It also seems good to take advantage of the numpy effort especially if it will ease some of the future burden for maintenance here.
> 
> For example: The `toarray()` function for COO nD arrays can be something like:
> 
> ```python
>     def toarray(self):
>         flat_indices = np.ravel_multi_index(self.coords, self.shape)
>         dense_result = np.zeros(self.shape, dtype=self.dtype)
>         dense_result.flat[flat_indices] = self.data
>         return dense_result
> ```


> If there are two approached already here (say for `def toarray`), it may be worth comparing those now to better understand the tradeoff @anushkasuyal.

Based on the two approaches, we could have the following definitions for `toarray`. (I've left the `self.ndim<3` case separate for now, to make sure it doesn't break the old code, but we could combine both if-else cases into one common case later on)

Here is the first approach, which uses coo_todense_nd:
```py
        else: # dim>2
            result = np.zeros(self.shape, dtype=self.dtype)
            coords = np.concatenate(self.coords)
            coo_todense_nd(self.shape, self.nnz, self.ndim,
                           coords, self.data, result.ravel('A'), fortran)
        return result.reshape(self.shape)
```

Here is the second approach, as proposed by @dschult, which uses numpytools:
```py
        else: # dim>2
            flat_indices = np.ravel_multi_index(self.coords, self.shape)
            result = np.zeros(self.shape, dtype=self.dtype)
            np.add.at(result.ravel(), flat_indices, self.data)
        return result.reshape(self.shape)
```
(I think there was a slight bug in this definition for when we had duplicate entries, but using `np.add.at` fixed that!)


I benchmarked the numpytools and coo_todense_nd approaches for `toarray`, for random sparse arrays with `density=0.1, dtype=int`, and here are the results I got -

| Shape               | Iterations | numpytools  | coo_todense_nd  |
|:--------------------|:-----------|:------------|:----------------|
| (100,100,100)       | 1000       | 3.75s, 4.27s | 3.45s, 3.33s     |
| (1000,100,1000)     | 10         | 4.35s, 4.08s | 5.17s, 6.09s     |
| (100,100,100,100)   | 10         | 5.20s, 4.91s | 7.93s, 7.96s     |
| (100,100,10,10,100) | 10         | 4.48s       | 8.33s           |

I think numpytools definitely outperforms `coo_todense_nd`, so this replacement seems like a sensible choice. I did have a question - in that case would we eliminate the pre-existing compiled code for `coo_todense` too?

**Comment by dschult** [comment]
Thanks for finding and fixing the bug with `np.add.at`. :)

> I think numpytools definitely outperforms coo_todense_nd, so this replacement seems like a sensible choice. I did have a question - in that case would we eliminate the pre-existing compiled code for coo_todense too?

We might want to try speeding up `coo_todense_nd` before replacing it. But I'm hoping that numpy is quick enough that we can avoid some new sparsetools functions.

As for `coo_todense`, we should test that code before we do anything with it. :)  Just because the nd version is slower doesn't mean the 2d is slower. But I think the bigger question is whether to support nD-arrays using a 2D representation of them. This is only one function among the many we would be rewriting in sparsetools as we build support for nD. But if the pattern holds, we can get pretty far by using numpy to work with 2D representations of the nD. And the existing 2D sparsetools can handle the rest. I am still wrapping my head around matmul in nD and how that might work with 2D representations of nD. But for most other functionality it seems like a promising approach.

**Comment by anushkasuyal** [comment]
> As for `coo_todense`, we should test that code before we do anything with it. :) Just because the nd version is slower doesn't mean the 2d is slower.

On benchmarking the densification of 2-D COO arrays (with `density=0.1`) using the numpytools and `coo_todense` approach, I got the following results -

| Shape         | Iterations | coo_todense  | numpytools   |
|:--------------|:-----------|:-------------|:-------------|
| (10,10)       | 10000      | 50ms, 44ms   | 47.6ms, 41ms |
| (100,100)     | 10000      | 78ms, 68ms   | 87ms, 94ms   |
| (1000,1000)   | 10000      | 11.7s, 11.4s | 41s, 45s     |
| (10000,10000) | 10         | 8.21s, 7.93s | 7.97s, 8.55s |

I think compiled code might actually be faster than NumPy for 2-D. Even for higher densities, `coo_todense` seemed to marginally outperform NumPy.

**Comment by dschult** [comment]
It is not surprising that CSR matmul is faster even with the conversion, because CSR was designed for sparse @ dense. Of course CSR @ CSR is not CSR @ dense. But it is still pretty good. Also this code is doing the conversion to CSR -- it sorts the indices and finds the start/stop pointers to each row. That's basically all the conversion `coo.tocsr()` does. :)

Some helpful idioms for this kind of work: 
To sort `A.coords`:  `A.sort_indices()`
To find when the rows change :  `indptr_row = np.diff(self.coords[0]).nonzero()[0]`
To find the nnz along each row: `nnz_by_row = np.diff(indptr)`

Another strategy that might be good to keep in mind (used by some sparsetools code): You can avoid sorting if you know the possible coords combinations that will give results. You set up a `result_data` array of zeros to hold all the resulting values and as you loop over the indices of both `A` and `B`, any match leads to a product that you add to the correct spot in the `result_data` array. I haven't worked through how that exactly fits with the rest of `matmul` for coo, but I believe it would. One possibly helpful function for finding matches could be `np.intersect1d`. But again, I haven't worked through a numpy tools version of matmul for coo. 

Since you are working on coo@coo, you should make efforts to stay away from finding the start/end of coords for each row.  That is the work of converting to csr (and the current code does that already) so you might as well convert to CSR and run that code. For nD-dot, I suggest figuring out which axis is being reduced (the inner axes for matmul) and raveling the other axes coords to get a 2D version of the nD. Do the same for the second array and use the current sparsetools 2D csr matmul code on the result. That might be a good competitor to C++ code written explicitly for nD. And it would be easier to maintain.  If the C++ code is faster, great! That's the reason for writing it. :)

**Comment by anushkasuyal** [comment]
I have collated the methods for COO arrays implemented across `_coo_base`, `_spbase`, and `_data_matrix` classes. We can use this to track what all is being implemented in each PR, with a brief note on its implementation.

The following is the current 1-D and 2-D COO array implementation in SciPy.

| **Methods**                              | **1- and 2-D** | **n-D** | **Implementation**                                      |
| ------------------------------------ | ---------- | --- | --------------------------------------------------- |
| `reshape`                              | Y          |    | in python                                       |
| `count_nonzero`                        | Y          |    | in python                                       |
| `transpose`                            | Y          |    | in python                                       |
| `resize`                               | Y          |     | in python                                           |
| `toarray`                              | Y          |    | 1- and 2-D use `coo_todense`|
| `tocsc`                                | Y          | \-  |                                                     |
| `tocsr`                                | Y          | \-  |                                                     |
| `tocoo`                                | Y          |    |                                                     |
| `todia`                                | Y          | \-  |                                                     |
| `todok`                                | Y          | \-  |                                                     |
| `diagonal`                             | Y          |     | in python                                           |
| `sum_duplicates`                       | Y          |    | in python                                       |
| `eliminate_zeros`                      | Y          |    | in python                                       |
| `_add_dense`                           | Y          |    | 1- and 2-D use `coo_todense` |
| `_add_sparse`                          | Y          |    | 1- and 2-D route via CSR          |
| `_sub_dense`                           | Y          |    | convert to dense and subtract                 |
| `_sub_sparse`                          | Y          |   | 1- and 2-D route via CSR         |
| ` _matmul_vector` (sparse-dense)     | Y         |    | 1- and 2-D use `coo_matvec`  |
| `_matmul_multivector` (sparse-dense) | Y          |     | 1- and 2-D use `coo_matmat_dense`                     |
| `_matmul_sparse` (sparse-sparse)     | Y          |     | 1- and 2-D route via CSR                            |
| `multiply`                             | Y          |     | via CSR                                             |
| `maximum`                              | Y          |     | via CSR                                             |
| `minimum`                              | Y          |     | via CSR                                             |
| `max`                                  | Y          |     |implemented in `_data_matrix`  |
| `min`                                  | Y          |     | implemented in `_data_matrix`  |
| `nanmax`                               | Y          |     | implemented in `_data_matrix`  |
| `nanmin`                               | Y          |     | implemented in `_data_matrix` |
| `power`                                | Y          |    | implemented in `_data_matrix`  |
| `eq`                                   | Y          |     | via CSR                                             |
| `ne`                                   | Y          |     | via CSR                                             |
| `lt`                                   | Y          |     | via CSR                                             |
| `gt`                                   | Y          |     | via CSR                                             |
| `le`                                   | Y          |     | via CSR                                             |
| `ge`                                   | Y          |     | via CSR                                             |
| `abs`                                  | Y          |   | uses built-ins in `_data_matrix`                     |
| `round`                                | Y          |    | uses np.around in `_data_matrix`                     |
| `div`                                  | Y          |     | uses numpy and routes via CSR                       |
| `dot` (sparse-dense)                   | Y          |     | 1- and 2-D uses matmul                              |
| `dot` (sparse-sparse)                 | Y          |     | 1- and 2-D uses matmul                              |
| `tensordot` (sparse-dense)             |            |     |                                                     |
| `tensordot` (sparse-sparse)            |            |     |                                                     |

**Comment by dschult** [comment]
Did we switch the `toarray` method for 1D input to use coo_todense_nd?

Also, what tool were you using for benchmarking/timing the methods? Is it a short script you could post somewhere?
Would it be easy to rerun the coo_todense_nd vs numpytools comparison now that the input is strides instead of shape? Don't spend a lot of time on this -- I can create something fairly quickly and try it. Just thought it might be easy.

**Comment by anushkasuyal** [comment]
> Did we switch the toarray method for 1D input to use coo_todense_nd?

Yes! I think CJ pointed out that `coo_todense` took an empty row array so using `coo_todense_nd` would avoid that.

>Also, what tool were you using for benchmarking/timing the methods? Is it a short script you could post somewhere?

I had asv installed and just used `python dev.py bench ...`. It was a short script I wrote then, I'll rewrite and post it.

Also, after I had already made the first PR, while working on the third chunk of work (which will follow the matmul PR) I ended up switching from `coo_todense_nd` to the numpy tools definition of `toarray()` from your POC. While working on the first PR, I wrote `coo_todense_nd` anyway, for dense add/sub. But in the third PR (yet to be made), I added broadcasting for add, sub, and all other element-wise operations, like we have in `np.add` etc. Then I just broadcasted, mapped n-D to 2-D, and routed via 2-D CSR addition, so `coo_todense_nd` could be eliminated fully.
So basically, if all the work eventually ends up going in, it's possible that the `coo_todense_nd` that was added might just be removed altogether later on.

**Comment by anushkasuyal** [comment]
@dschult I reran the benchmarks to compare, and got the following results (with `density=0.1`) -
| Shape               | Iterations | numpytools  | coo_todense_nd  |
|:--------------------|:-----------|:------------|:----------------|
| (100,100,100)       | 1000       | 2.26s, 2.33s | 2.58s, 2.50s     |
| (1000,100,1000)     | 10         | 3.11s, 3.54s | 3.91s, 4.34s     |
| (100,100,100,100)   | 10         | 3.51s, 2.89s | 5.47s, 5.33s     |
| (100,100,10,10,100) | 10         | 3.55s, 3.33s  | 6.54s, 6.98s    |

The script is quite short -
```py3
class toarray(Benchmark):
    def setup(self):
        self.arr = sparse.random_array(shape=(100,100,10,10,100), density=0.1, dtype=int)

    def time_large(self):
        for i in range(10):
            self.arr.toarray()
```
I just kept switching between the two `toarray()` definitions while benchmarking for `coo_todense_nd` and `numpytools`.

**Comment by anushkasuyal** [comment]
After the approval of the first (#21197) and second (#21435) PRs that work towards extending nD support for COOrdinate sparse arrays in SciPy, the method tracker post updates is as follows -

| **Methods**                              | **1D/2D** | **nD** | **Implementation**                                      |
| ------------------------------------ | ---------- | --- | --------------------------------------------------- |
| `reshape`                              | Y          | Y   | all-D in python                                   |
| `count_nonzero`                        | Y          | Y   | all-D in python                                   |
| `transpose`                            | Y          | Y   | all-D in python                                   |
| `resize`                               | Y          |     | 1- and 2-D in python                               |
| `toarray`                              | Y          | Y   | 2-D uses `coo_todense`; 1- and n-D use `coo_todense_nd` |
| `tocsc`                                | Y          | \-  |                                                     |
| `tocsr`                                | Y          | \-  |                                                     |
| `tocoo`                                | Y          | Y   |                                                     |
| `todia`                                | Y          | \-  |                                                     |
| `todok`                                | Y          | \-  |                                                     |
| `diagonal`                             | Y          |     | in python                                           |
| `sum_duplicates`                       | Y          | Y   | all-D in python                                       |
| `eliminate_zeros`                      | Y          | Y   | all-D in python                                       |
| `_add_dense`                           | Y          | Y   | 1- and 2-D use `coo_todense`; n-D uses `coo_todense_nd` |
| `_add_sparse`                          | Y          | Y   | 1- and 2-D route via CSR; n-D uses python           |
| `_sub_dense`                           | Y          | Y   | all-D convert to dense and subtract                 |
| `_sub_sparse`                          | Y          | Y   | 1- and 2-D route via CSR; n-D uses python           |
| ` _matmul_vector` (sparse-dense)     | Y          | Y   | 1- and 2-D use `coo_matvec`; n-D uses `coo_matvec_nd`   |
| `_matmul_multivector` (sparse-dense) | Y          | Y | 1- and 2-D use `coo_matmat_dense`; n-D uses `coo-matmat-dense-nd` |
| `_matmul_sparse` (sparse-sparse)     | Y          |     | 1- and 2-D route via CSR; n-D maps to 2-D then routes via CSR         |
| `multiply`                             | Y          |     | via CSR                                             |
| `maximum`                              | Y          |     | via CSR                                             |
| `minimum`                              | Y          |     | via CSR                                             |
| `max`                                  | Y          |     |implemented upto 2-D in `_data_matrix`  |
| `min`                                  | Y          |     | implemented upto 2-D in `_data_matrix`  |
| `nanmax`                               | Y          |     | implemented upto 2-D in `_data_matrix`  |
| `nanmin`                               | Y          |     | implemented upto 2-D in `_data_matrix` |
| `power`                                | Y          | Y   | implemented for all-D in `_data_matrix`  |
| `eq`                                   | Y          |     | via CSR                                             |
| `ne`                                   | Y          |     | via CSR                                             |
| `lt`                                   | Y          |     | via CSR                                             |
| `gt`                                   | Y          |     | via CSR                                             |
| `le`                                   | Y          |     | via CSR                                             |
| `ge`                                   | Y          |     | via CSR                                             |
| `abs`                                  | Y          | Y   | uses built-ins in `_data_matrix`                     |
| `round`                                | Y          | Y   | uses np.around in `_data_matrix`                     |
| `div`                                  | Y          |     | uses numpy and routes via CSR                       |
| `dot` (sparse-dense)                   | Y          | Y  | 1- and 2-D uses matmul; map n-D to 2-D and use `csr_matmat` |
| `dot` (sparse-sparse)                 | Y          | Y  | 1- and 2-D uses matmul; reshape then use n-D `matmul` |
| `tensordot` (sparse-dense)             |   Y     |  Y   |   map n-D to 2-D and use `csr_matmat`        |
| `tensordot` (sparse-sparse)            |  Y      |  Y   |   reshape then use n-D `matmul`        |
| `broadcast_to`                    |  Y      |  Y   |   all-D use numpytools                |
| `todense`                          | Y            | Y    |                                                       |
| `random_array`                          | Y            | Y    |                                                       |
| `hstack`                          |  Y           |     |                                                       |
| `vstack`                          |  Y           |     |                                                       |
| `block_array`                          | Y            |     |                                                       |
| `sum`                          |  Y           |     |                                                       |
| `mean`                          | Y            |     |                                                       |
| `triu`                          | Y            |     |                                                       |
| `tril`                          | Y            |     |                                                       |

**Comment by anushkasuyal** [comment]
The remaining work will be divided into 2 more PRs to ease reviewing (as a lot of new code has been introduced).

The 3rd follow-up PR will extend operations that are implemented using the following set of steps - broadcasting (in the case of element-wise operations), mapping to 2-D, routing via 2-D CSR, and mapping back. This will involve methods such as boolean comparators (`<`, `>`, `==` etc.), `multiply`, `divide`, `maximum`, `minimum`, `max`, `min`, `sum`, `mean` etc.

The 4th follow-up PR will then extend all remaining methods to n-D, including some linear algebra operations such as `diagonal`, `tensorsolve`, `kron`, `triu`, `tril` etc. and constructors such as `vstack`, `hstack`, and `block_array`.

**Comment by anushkasuyal** [comment]
The third follow-up PR working towards extending n-D support for COO arrays has been made at https://github.com/scipy/scipy/pull/21613.
